File: src/ai/providers/gemini.ts
/**
 * Google Gemini provider adapter.
 * Supports Gemini 2.5 Flash and Flash Lite models.
 *
 * Endpoint: https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent
 *           https://generativelanguage.googleapis.com/v1beta/models/{model}:streamGenerateContent
 * Auth:     ?key= query parameter
 * Messages: contents[].parts[], role "model" instead of "assistant"
 * System:   systemInstruction top-level field
 * Thinking: thinkingConfig.thinkingBudget for thinking-capable models
 */

import type { ChatMessage, ChatOptions, StreamEvent } from "../types";
import { BaseProvider } from "./base";

const BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models";

interface GeminiPart {
  text?: string;
  inlineData?: {
    mimeType: string;
    data: string;
  };
  thought?: boolean;
}

interface GeminiContent {
  role: string;
  parts: GeminiPart[];
}

interface GeminiCandidate {
  content: {
    parts: GeminiPart[];
    role: string;
  };
}

export class GeminiProvider extends BaseProvider {
  async chat(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): Promise<ChatMessage> {
    const url = `${BASE_URL}/${model}:generateContent?key=${this.apiKey}`;
    const body = this.buildRequestBody(messages, options);

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(
        `Gemini request failed (${response.status}): ${errorText}`
      );
    }

    const data = await response.json();
    const candidates = data.candidates as GeminiCandidate[] | undefined;

    let content = "";
    let thinking = "";

    if (candidates && candidates.length > 0) {
      const parts = candidates[0].content?.parts ?? [];
      for (const part of parts) {
        if (part.thought && part.text) {
          thinking += part.text;
        } else if (part.text) {
          content += part.text;
        }
      }
    }

    return {
      role: "assistant",
      content,
      thinking: thinking || undefined,
      timestamp: Date.now(),
    };
  }

  async *chatStream(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): AsyncGenerator<StreamEvent> {
    const url = `${BASE_URL}/${model}:streamGenerateContent?alt=sse&key=${this.apiKey}`;
    const body = this.buildRequestBody(messages, options);

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      yield {
        type: "error",
        message: `Gemini stream failed (${response.status}): ${errorText}`,
      };
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) {
      yield { type: "error", message: "No response body from Gemini" };
      return;
    }

    let inputTokens = 0;
    let outputTokens = 0;

    for await (const chunk of this.parseSSE(reader)) {
      const data = chunk as Record<string, unknown>;

      // Extract usage metadata
      if (data.usageMetadata) {
        const usage = data.usageMetadata as Record<string, number>;
        inputTokens = usage.promptTokenCount ?? inputTokens;
        outputTokens = usage.candidatesTokenCount ?? outputTokens;
      }

      const candidates = data.candidates as GeminiCandidate[] | undefined;
      if (!candidates || candidates.length === 0) continue;

      const parts = candidates[0].content?.parts;
      if (!parts) continue;

      for (const part of parts) {
        if (part.thought && part.text) {
          yield { type: "thinking", content: part.text };
        } else if (part.text) {
          yield { type: "content", content: part.text };
        }
      }
    }

    yield {
      type: "done",
      usage: { input_tokens: inputTokens, output_tokens: outputTokens },
    };
  }

  private buildRequestBody(
    messages: ChatMessage[],
    options: ChatOptions | undefined
  ): Record<string, unknown> {
    // Extract system messages for systemInstruction
    const systemMessages = messages.filter((m) => m.role === "system");
    const nonSystemMessages = messages.filter((m) => m.role !== "system");

    const body: Record<string, unknown> = {
      contents: this.formatContents(nonSystemMessages),
    };

    // System prompt as systemInstruction
    if (systemMessages.length > 0) {
      body.systemInstruction = {
        parts: [{ text: systemMessages.map((m) => m.content).join("\n\n") }],
      };
    }

    // Generation config
    const generationConfig: Record<string, unknown> = {};

    if (options?.maxTokens) {
      generationConfig.maxOutputTokens = options.maxTokens;
    }

    if (options?.temperature !== undefined) {
      generationConfig.temperature = options.temperature;
    }

    // Thinking configuration for thinking-capable models
    if (options?.thinking) {
      generationConfig.thinkingConfig = {
        thinkingBudget: options.maxTokens
          ? Math.floor(options.maxTokens * 0.8)
          : 10000,
      };
    }

    if (Object.keys(generationConfig).length > 0) {
      body.generationConfig = generationConfig;
    }

    return body;
  }

  private formatContents(messages: ChatMessage[]): GeminiContent[] {
    return messages.map((m) => {
      // Gemini uses "model" instead of "assistant"
      const role = m.role === "assistant" ? "model" : "user";

      const parts: GeminiPart[] = [];

      // Add images first if present
      if (m.images && m.images.length > 0) {
        for (const img of m.images) {
          let mimeType = "image/jpeg";
          let data = img;

          const dataUriMatch = img.match(
            /^data:(image\/[a-z+]+);base64,(.+)$/
          );
          if (dataUriMatch) {
            mimeType = dataUriMatch[1];
            data = dataUriMatch[2];
          }

          parts.push({
            inlineData: { mimeType, data },
          });
        }
      }

      parts.push({ text: m.content });

      return { role, parts };
    });
  }
}

File: crates/provider-adapters/src/gemini.rs
use serde_json::{json, Value};

use super::types::{ChatRequest, ChatResponse, StreamEvent, Usage};

const GEMINI_API_URL: &str = "https://generativelanguage.googleapis.com/v1beta/models";

/// Build the request URL for Gemini. The API key is passed as a query parameter.
fn build_url(model: &str, api_key: &str, stream: bool) -> String {
    let action = if stream {
        "streamGenerateContent?alt=sse"
    } else {
        "generateContent"
    };
    format!(
        "{}/{}:{}{}key={}",
        GEMINI_API_URL,
        model,
        action,
        if stream { "&" } else { "?" },
        api_key
    )
}

/// Build the Gemini API request body.
///
/// Gemini uses `contents[].parts[]` for messages and `systemInstruction` for
/// system prompts. Role names differ: "user" stays "user", but "assistant"
/// becomes "model".
fn build_request_body(request: &ChatRequest) -> Value {
    let mut system_text: Option<String> = None;
    let mut contents: Vec<Value> = Vec::new();

    for msg in &request.messages {
        if msg.role == "system" {
            system_text = Some(msg.content.clone());
        } else {
            let role = match msg.role.as_str() {
                "assistant" => "model",
                other => other,
            };
            contents.push(json!({
                "role": role,
                "parts": [{ "text": msg.content }]
            }));
        }
    }

    let mut body = json!({
        "contents": contents,
        "generationConfig": {
            "maxOutputTokens": request.max_tokens,
            "temperature": request.temperature,
        }
    });

    if let Some(sys) = system_text {
        body["systemInstruction"] = json!({
            "parts": [{ "text": sys }]
        });
    }

    body
}

/// Send a non-streaming request to the Gemini API.
pub async fn send_request(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<ChatResponse, String> {
    let url = build_url(&request.model, api_key, false);
    let body = build_request_body(request);

    let response = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Gemini request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!("Gemini API error ({}): {}", status, text));
    }

    let data: Value = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse Gemini response: {}", e))?;

    let content = data["candidates"][0]["content"]["parts"][0]["text"]
        .as_str()
        .unwrap_or("")
        .to_string();

    let input_tokens = data["usageMetadata"]["promptTokenCount"]
        .as_u64()
        .unwrap_or(0) as u32;
    let output_tokens = data["usageMetadata"]["candidatesTokenCount"]
        .as_u64()
        .unwrap_or(0) as u32;

    Ok(ChatResponse {
        content,
        thinking: None,
        usage: Usage {
            input_tokens,
            output_tokens,
        },
    })
}

/// Send a streaming request to the Gemini API.
pub async fn send_stream(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<reqwest::Response, String> {
    let url = build_url(&request.model, api_key, true);
    let body = build_request_body(request);

    let response = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Gemini stream request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!("Gemini stream API error ({}): {}", status, text));
    }

    Ok(response)
}

/// Parse an SSE chunk from Gemini's streaming format into unified StreamEvents.
///
/// Gemini streaming returns `data: <json>` lines where each JSON object
/// contains `candidates[0].content.parts[0].text` for content deltas.
pub fn parse_stream_chunk(chunk: &str) -> Vec<StreamEvent> {
    let mut events = Vec::new();

    for line in chunk.lines() {
        let line = line.trim();
        if !line.starts_with("data: ") {
            continue;
        }
        let data = &line[6..];

        if let Ok(parsed) = serde_json::from_str::<Value>(data) {
            // Extract text content
            if let Some(text) = parsed["candidates"][0]["content"]["parts"][0]["text"].as_str() {
                if !text.is_empty() {
                    events.push(StreamEvent {
                        event_type: "content".to_string(),
                        content: Some(text.to_string()),
                        usage: None,
                        message: None,
                    });
                }
            }

            // Extract usage metadata
            if let Some(usage_meta) = parsed.get("usageMetadata") {
                if !usage_meta.is_null() {
                    let input_tokens =
                        usage_meta["promptTokenCount"].as_u64().unwrap_or(0) as u32;
                    let output_tokens =
                        usage_meta["candidatesTokenCount"].as_u64().unwrap_or(0) as u32;
                    if input_tokens > 0 || output_tokens > 0 {
                        events.push(StreamEvent {
                            event_type: "usage".to_string(),
                            content: None,
                            usage: Some(Usage {
                                input_tokens,
                                output_tokens,
                            }),
                            message: None,
                        });
                    }
                }
            }

            // Check for finish reason
            if let Some(reason) = parsed["candidates"][0]["finishReason"].as_str() {
                if reason == "STOP" || reason == "MAX_TOKENS" {
                    events.push(StreamEvent {
                        event_type: "done".to_string(),
                        content: None,
                        usage: None,
                        message: None,
                    });
                }
            }
        }
    }

    events
}
